invenio_base/urls/builder.py

```diff
@@ -0,0 +1,130 @@
+to Flask's `url_for` which allows any part of the application (in the broad
+sense of the term) to generate URLs even if those URLs are for views registered
+in another Flask application. Other developer niceties are included.
+"""
+
+from abc import ABC, abstractmethod
+
+from flask import Flask, current_app
+from werkzeug.routing import BuildError, Map, Rule
+
+from invenio_base.app import blueprint_loader
+
+
+class InvenioUrlsBuilder(ABC):
+    """Interface of class in charge of producing urls."""
+
+    @abstractmethod
+    def build(self, endpoint, values, method=None):
+        """Build current or other app url."""
+
+
+class InvenioAppsUrlsBuilder(InvenioUrlsBuilder):
+    """Builds URLs with some knowledge of Invenio (app-rdm)."""
+
+    def __init__(
+        self,
+        cfg_of_app_prefix,
+        cfg_of_other_app_prefix,
+        groups_of_other_app_entrypoints,
+    ):
+        """Constructor."""
+        self.cfg_of_app_prefix = cfg_of_app_prefix
+        self.cfg_of_other_app_prefix = cfg_of_other_app_prefix
+        self.groups_of_other_app_entrypoints = groups_of_other_app_entrypoints
+
+    def setup(self, app, **kwargs):
+        """Sets up the object for url generation.
+
+        It does so by building an internal url_map that it will reuse.
+
+        This is called before the application is fully setup (not in an application
+        context).
+        """
+        # Create a tmp Flask app
+        app_tmp = Flask("InvenioAppsUrlsBuilder")
+        app_tmp.config = app.config
+
+        blueprint_loader(
+            app_tmp,
+            entry_points=self.groups_of_other_app_entrypoints,
+        )
+
+        # Copy the Rules minus the view_functions (don't need them)
+        self.url_map = Map(
+            [Rule(r.rule, endpoint=r.endpoint) for r in app_tmp.url_map.iter_rules()]
+        )
+
+    def prefix(self, site_cfg):
+        """Return site prefix."""
+        return current_app.config[site_cfg]
+
+    def build(self, endpoint, values, method=None):
+        """Build full url of any registered endpoint with appropriate prefix.
+
+        This is called within an application context.
+        """
+        # 1- Try to build url from current app
+        try:
+            # discuss: TODO: Consider if cache on application context (g)
+            url_adapter = current_app.url_map.bind("")
+            url_relative = url_adapter.build(
+                endpoint, values, method=method, force_external=False
+            )
+            return self.prefix(self.cfg_of_app_prefix) + url_relative
+        except BuildError:
+            # The endpoint may be from the complementary blueprints
+            pass
+
+        # 2- Try to build url from complementary url_map
+        # discuss: TODO: Consider caching on application context (g)
+        url_adapter = self.url_map.bind("")
+        url_relative = url_adapter.build(  # type: ignore[union-attr]
```

Comment:
minor: lifted from original `url_for` which had typing. I should be able to remove that comment.